File: java/src/main/java/it/gabrieletondi/telldontaskkata/usecase/OrderCreationUseCase.java
package it.gabrieletondi.telldontaskkata.usecase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderBuilder;
import it.gabrieletondi.telldontaskkata.domain.OrderItem;
import it.gabrieletondi.telldontaskkata.domain.OrderItemBuilder;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.domain.Product;
import it.gabrieletondi.telldontaskkata.repository.OrderRepository;
import it.gabrieletondi.telldontaskkata.repository.ProductCatalog;
import it.gabrieletondi.telldontaskkata.usecase.exception.UnknownProductException;

public class OrderCreationUseCase {
    private final OrderRepository orderRepository;
    private final ProductCatalog productCatalog;

    public OrderCreationUseCase(OrderRepository orderRepository, ProductCatalog productCatalog) {
        this.orderRepository = orderRepository;
        this.productCatalog = productCatalog;
    }

    public void run(SellItemsRequest request) {
        Order order = new OrderBuilder()
                .withStatus(OrderStatus.CREATED)
                .withCurrency("EUR")
                .createOrder();

        for (SellItemRequest itemRequest : request.requests()) {
            Product product = productCatalog.getByName(itemRequest.productName());

            if (product == null) {
                throw new UnknownProductException();
            }

            final OrderItem orderItem = new OrderItemBuilder()
                    .withProduct(product)
                    .withQuantity(itemRequest.quantity())
                    .withTax(product.getTaxAmount(itemRequest.quantity()))
                    .withTaxedAmount(product.getTaxedAmount(itemRequest.quantity()))
                    .createOrderItem();
            order.addItem(orderItem);
        }

        orderRepository.save(order);
    }

}

File: typescript/src/useCase/OrderCreationUseCase.ts
import Order from '../domain/Order';
import OrderItem from '../domain/OrderItem';
import {OrderStatus} from '../domain/OrderStatus';
import Product from '../domain/Product';
import OrderRepository from '../repository/OrderRepository';
import {ProductCatalog} from '../repository/ProductCatalog';
import SellItemsRequest from './SellItemsRequest';
import UnknownProductException from './UnknownProductException';

class OrderCreationUseCase {
  private readonly orderRepository: OrderRepository;
  private readonly productCatalog: ProductCatalog;

  public constructor(orderRepository: OrderRepository, productCatalog: ProductCatalog) {
    this.orderRepository = orderRepository;
    this.productCatalog = productCatalog;
  }

  public run(request: SellItemsRequest): void {
    const order: Order = new Order(OrderStatus.CREATED, 1, 'EUR');

    for (const itemRequest of request.getRequests()) {
      const product: Product = this.productCatalog.getByName(itemRequest.getProductName());

      if (product === undefined) {
        throw new UnknownProductException();
      }
      else {
        const taxedAmount: number = product.calculateTaxedAmount(itemRequest.getQuantity());
        const taxAmount: number = product.calculateTaxAmount(itemRequest.getQuantity());

        const orderItem: OrderItem = new OrderItem(product, itemRequest.getQuantity(), taxAmount, taxedAmount);
        order.addOrderItem(orderItem);
      }
    }

    this.orderRepository.save(order);
  }
}

export default OrderCreationUseCase;
